src/renderers/testing/ReactShallowRenderer.js

```diff
@@ -100,26 +59,146 @@ class ReactShallowRenderer {
+
+          checkPropTypes(
+            element.type.contextTypes,
+            context,
+            'context',
+            getName(element.type, this._instance),
+            ReactDebugCurrentFrame.getStackAddendum,
+          );
+
+          ReactDebugCurrentFrame.element = null;
+        }
+
+        this._rendered = mountClassComponent(
+          this._instance,
+          element.props,
+          context,
+        );
+      } else {
+        this._rendered = element.type();
+      }
+    }
 
     return this.getRenderOutput();
   }
 
   unmount() {
-    this._renderer.unmount();
+    if (this._instance) {
+      if (typeof this._instance.componentWillUnmount === 'function') {
+        this._instance.componentWillUnmount();
+      }
+    }
+
+    this._element = null;
+    this._rendered = null;
+    this._instance = null;
   }
 }
 
-// Backwards compatible API
-ReactShallowRenderer.createRenderer = function() {
-  return new ReactShallowRenderer();
-};
+class Updater {
+  constructor(renderer) {
+    this._renderer = renderer;
+  }
+
+  isMounted(publicInstance) {
+    return !!this._renderer._element;
+  }
+
+  enqueueForceUpdate(publicInstance, callback, callerName) {
+    this._renderer.render(this._renderer._element); // TODO
+  }
+
+  enqueueReplaceState(publicInstance, completeState, callback, callerName) {
+    publicInstance.state = completeState;
+    this._renderer.render(this._renderer._element);
+  }
+
+  enqueueSetState(publicInstance, partialState, callback, callerName) {
+    publicInstance.state = {
+      ...publicInstance.state,
+      ...partialState,
+    };
+    this._renderer.render(this._renderer._element);
+  }
+}
+
+function getName(type, instance) {
+  var constructor = instance && instance.constructor;
+  return type.displayName ||
+    (constructor && constructor.displayName) ||
+    type.name ||
+    (constructor && constructor.name) ||
+    null;
+}
+
+function mountClassComponent(
+  instance,
+  props = emptyObject,
+  context = emptyObject,
+) {
+  instance.context = context;
+  instance.props = props;
+  instance.state = instance.state || emptyObject;
+
+  if (typeof instance.componentWillMount === 'function') {
+    instance.componentWillMount();
+  }
+
+  const rendered = instance.render();
+
+  if (typeof instance.componentDidMount === 'function') {
+    instance.componentDidMount();
+  }
+
+  return rendered;
+}
+
+function updateClassComponent(
+  instance,
+  props = emptyObject,
+  context = emptyObject,
+) {
+  const oldContext = instance.context;
+  const oldProps = instance.props;
+  const oldState = instance.state;
+
+  if (typeof instance.componentWillUpdate === 'function') {
+    instance.componentWillUpdate(props, instance.state, context);
+  }
+
+  instance.context = context;
+  instance.props = props;
+
+  const rendered = instance.render();
+
+  if (typeof instance.componentDidUpdate === 'function') {
+    instance.componentDidUpdate(oldProps, oldState, oldContext);
```

Comment:
@bvaughn 
I guess the reason is that `ShallowRender.render` run on an outside transaction, which means that enqueued functions have never been called.

* https://github.com/facebook/react/blob/master/src/test/ReactShallowRenderer.js#L68
* https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L1043-L1053

I used to create a PR to fix it. I hope it can be of any help to you.

* https://github.com/facebook/react/pull/7912